### PR TITLE
CI Pipeline

### PR DESCRIPTION
**Pull request by noahmacaskill**
Draft PR for setting up CI pipeline. Tests are running, but some are failing.

### CODE REVIEW

**Comment by atruskie** [comment]
Just a comment to begin with: you've based this branch off of commits that aren't in the main repo. Always update your forks before you start new work.

image

File: .github/workflows/dotnet.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+name: .NET
+
+on:
+  push:
+    branches: [ master, CIpipeline ]
```
Comment (atruskie) [inline .github/workflows/dotnet.yml:5]: We should run tests on every branch, not just `master` and `CIpipeline`.  Try:

```yaml
on:
  push:
    branches:
      - "*"
```

File: .github/workflows/dotnet.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  pull_request:
+    branches: [ master ]
```
Comment (atruskie) [inline .github/workflows/dotnet.yml:7]: same comment as above

File: .github/workflows/dotnet.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  build:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Setup .NET
+      uses: actions/setup-dotnet@v1
+      with:
+        dotnet-version: 6.0.x
```
Comment (atruskie) [inline .github/workflows/dotnet.yml:19]: You can omit the version, and the version installed will be based off of the `global.json` file: https://github.com/actions/setup-dotnet#usage

File: .github/workflows/dotnet.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+on:
+  push:
+    branches: [ master, CIpipeline ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
```
Comment (atruskie) [inline .github/workflows/dotnet.yml:12]: It'd be good to see a matrix build here for windows, macos, and linux

File: test/MetadataUtility.Tests/TestHelpers/Extensions.cs [hunk 5]
```diff
@@ -5,12 +5,18 @@
...
+    using System.Text.RegularExpressions;
 
     public static class Extensions
     {
         public static void AddEmptyFile(this MockFileSystem fileSystem, string path)
         {
             fileSystem.AddFile(path, string.Empty);
         }
+
+        public static string toCrLf(string str)
```
Comment (atruskie) [inline test/MetadataUtility.Tests/TestHelpers/Extensions.cs:17]: Methods have a leading capital in .NET

**Review by atruskie — CHANGES_REQUESTED** [review]

**Comment by atruskie** [comment]
Ok, since the tests could keep failing for a while, I'd rather see the main repo start building on changes. 

So make the changes suggested above, then we'll merge, and deal with tests in another edit

**Comment by noahmacaskill** [comment]
I think everything's been addressed. Moved some things around with some rebasing to clean those old commits out of this fork. Made a super basic matrix build for running on windows, macos and ubuntu, noticed AP had a lot more detail here, can bring some more of that over if needed. Since some of our tests are still failing, the pipeline was halting execution as soon as one failed, to alleviate this I threw in a "continue-on-error" flag so that they would all run.

Another interesting thing I noticed, every once and a while the tests crash before they've all been run (upon a rerun it will usually run all the way through though I haven't changed anything). Error message:

Aborted tests

Will see if I can figure this out. If there's any other feedback, let me know!

**Comment by noahmacaskill** [comment]
Think I've solved the issue of the tests crashing, seems the issue was coming from [here](https://github.com/QutEcoacoustics/emu/blob/43354c0db8c77ced73be69316bf82243dd2a9efe/test/MetadataUtility.Tests/TestHelpers/FixtureHelper.cs#L109). For some reason this was sometimes trying to delete a temporary directory that doesn't exist (anymore at least). Though that seems to fix it, I'm not sure what was causing it to happen.

**Comment by atruskie** [comment]
@noahmacaskill that's good information. `TestTempDir` is only used in one spot: https://github.com/QutEcoacoustics/emu/blob/8244ad941296ba3ad01f1fc60b370fa45d3dabf4/test/MetadataUtility.Tests/Utilities/FileMatcherTests.cs#L67

I think given the type of code involved here, we'll just add a try/catch around that directory deletion.

**Comment by atruskie** [comment]
(Sorry - fat-thumbed it)

Okay time to merge. I'd like you to squash all your commits in this branch into one. Then I'll merge it.

https://stackoverflow.com/questions/43978823/how-to-squash-commits-in-one-branch

**Comment by atruskie** [comment]
(Sorry - fat-thumbed it)

Okay time to merge. I'd like you to squash all your commits in this branch into one. Then I'll merge it.
https://stackoverflow.com/questions/43978823/how-to-squash-commits-in-one-branch

**Comment by noahmacaskill** [comment]
Looks like the same error was being encountered [here](https://github.com/QutEcoacoustics/emu/blob/43354c0db8c77ced73be69316bf82243dd2a9efe/test/MetadataUtility.Tests/TestHelpers/TempFile.cs#L54-L65), followed the same logic for this try/catch.

**Comment by noahmacaskill** [comment]
Squashed into one commit

**Pull request merged**